timeboost-sequencer/src/decrypt.rs

```diff
@@ -282,272 +208,318 @@ impl Worker {
-                                    warn!(node = %self.label, %err, "failed to insert local shares");
-                                    continue;
-                                }
-                                if catching_up {
-                                    // fast-forward
-                                    self.shares.retain(|k, _| {
-                                        let old = k.round() < round;
-                                        if old {
-                                            hatched_rounds.insert(k.round());
-                                        }
-                                        !old
-                                    });
-                                    catching_up = false;
-                                }
-                            }
-                            Err(err) => {
-                                warn!(node = %self.label, %err, "failed to decrypt data");
-                                continue;
-                            }
-                        }
-                    }
-                    Some(WorkerCommand::Gc(round)) => {
-                        let round = round.saturating_sub(MAX_ROUNDS as u64);
-                        // gc only if activity in the incubator.
-                        if round > 0 && !r.is_genesis() {
-                            trace!(node = %self.label, round, "gc");
-                            self.net.gc(round);
-                            hatched_rounds.retain(|r| round <= **r);
-                            continue;
-                        }
-                    }
-                    None => {
-                        debug!(node = %self.label, "decrypter shutdown");
-                        return;
-                    }
-                },
+            // workers don't have to hatch for unrequested rounds.
+            if self.first_requested_round.is_none() {
+                continue;
             }
 
-            // check for hatched ciphertexts
-            match self.hatch(r) {
-                Ok(Some((dec_round, dec_items))) => {
-                    if let Err(err) = dec_tx.send(WorkerResponse(dec_round, dec_items)).await {
-                        error!(node = %self.label, %err, "failed to send decrypted data");
-                        return;
+            // when processing the incoming event (all steps above), we already early-return and continue the loop if any intermediate steps error.
+            // If they succeed, then some local decryption shares must have been updated, thus we try to hatch them.
+            let round = self.oldest_cached_round();
+            match self.hatch(round) {
+                Ok(Some(incl)) => {
+                    while let Err(err) = res_tx.send(WorkerResponse::Decrypt(incl.clone())).await {
+                        error!(%node, %err, "failed to send hatched inclusion list, retrying");
+                        sleep(Duration::from_millis(50)).await;
```

Comment:
I don't think this error is recoverable. Once the receiver is dropped or has closed the channel the worker can not make progress and should likewise terminate execution.